src/puppetlabs/puppetdb/command.clj

```diff
@@ -214,50 +223,37 @@
 
 ;; ## Command submission
 
-(defn-validated ^:private do-enqueue-raw-command :- s/Str
-  "Submits raw-command to the mq-endpoint of mq-connection and returns
-  its id."
-  [mq-connection :- mq/connection-schema
-   mq-endpoint :- s/Str
-   raw-command :- (s/cond-pre s/Str utils/byte-array-class)
-   uuid :- (s/maybe s/Str)
-   properties :- (s/maybe {s/Str s/Str})] ;; For now stick with str -> str
-  (let [uuid (or uuid (kitchensink/uuid))]
-    (mq/send-message! mq-connection mq-endpoint raw-command
-                      ;; Until/unless we require that all callers
-                      ;; include received, etc.
-                      (merge {"received" (kitchensink/timestamp) "id" uuid}
-                             properties))
-    uuid))
-
-(defn-validated ^:private do-enqueue-command :- s/Str
+(defn-validated do-enqueue-command
   "Submits command to the mq-endpoint of mq-connection and returns
   its id. Annotates the command via annotate-command."
-  [mq-connection :- mq/connection-schema
-   mq-endpoint :- s/Str
+  [q
+   command-chan
+   ^Semaphore write-semaphore
    command :- s/Str
    version :- s/Int
-   payload
-   uuid :- (s/maybe s/Str)
-   properties]
-  (let [command-map {:command command
-                     :version version
-                     :payload payload}]
-    (do-enqueue-raw-command mq-connection mq-endpoint
-                            (json/generate-string command-map)
-                            uuid
-                            properties)))
+   certname :- s/Str
+   command-stream
+   command-callback]
+  (try
+    (.acquire write-semaphore)
+    (time! (get @metrics :message-persistence-time)
+           (async/>!! command-chan
+                      (queue/store-command
+                        q command version certname command-stream command-callback)))
+    (finally
+      (.release write-semaphore)
+      (when command-stream
+        (.close command-stream)))))
```

Comment:
The caller in this case is Jetty, so it's already opened when we get it, I can add it to the docstring